### PR TITLE
Test: Implement auth contract test cases

### LINKED ISSUE #123: Implement comprehensive test cases for Auth Contract
# Test Coverage for Auth Contract

## Overview
Implement comprehensive test cases for the Auth Contract to ensure reliability and correctness.

## Test Scenarios to Cover
1. Account Creation
   - Successful account creation with valid parameters
   - Error cases with invalid parameters

2. Authorization Checks
   - Valid authorization scenarios
   - Invalid signature verification
   - Challenge verification
   - Device management authorization

3. Device Management
   - Adding devices
   - Removing devices
   - Device limit checks

4. Recovery Address Management
   - Adding recovery address
   - Updating recovery address
   - Recovery process testing

5. Error Cases
   - Invalid parameters
   - Unauthorized access attempts
   - Edge cases in signature verification

## References
- Related to PR: #120
- Discussion: https://github.com/kindfi-org/kindfi/pull/120#discussion_r1938939822

## Implementation Guidelines
- Use proper test fixtures and mocks where necessary
- Cover both success and failure scenarios
- Include edge cases and boundary conditions
- Follow Rust testing best practices

### PR DESCRIPTION
**Pull request by psychemist**
# Pull Request Overview

## Summary
This PR implements a comprehensive test suite for the KindFi authentication system, covering the AuthController, Account, and AccountFactory contracts, ensuring secure multi-signature authentication, account management, and factory operations with robust error handling and event verification.

### Related Issues
- Closes #123

### Type of Change
Mark with an `x` all the checkboxes that apply (like `[x]`).

- [ ] Feat (A new feature)
- [ ] Fix (A bug fix)
- [x] Docs (Documentation changes)
- [x] Style (changes that do not affect the meaning of the code: formatting, etc)
- [ ] Refactor (code changes that neither fix a bug nor add a feature)
- [x] Test (adding missing tests or correcting existing tests)
- [ ] Build (changes that affect the build system or external dependencies)
- [ ] Chore (maintenance changes that do not fall into any of the other categories)

### Changes Made
1. Created a comprehensive test suite for the KindFi authentication system covering:
   - AuthController, AccountFactory, and AccountContract in `test_auth.rs`
   - Unit and Integration tests demonstrating interactions between these components
   - Specialized testing utilities for cryptographic operations and signature verification
   - Test fixtures for streamlined setup and consistent test environments
   - Dedicated README.md with detailed documentation of the test structure and usage

2. Implemented extensive test scenarios covering:
   - Core functionality tests for all contracts (initialization, authentication, device management)
   - Edge case handling (authorization failures, threshold violations, duplicate signers)
   - Integration scenarios between all three contracts
   - Event emission verification for all contract operations
   - Security validations for the entire authentication flow

### Technical Notes
1. **Testing Architecture**:
   - Implemented the `SecureKeyStorage` utility in AuthController's test_auth.rs for secure key management
   - Created specialized deterministic signature generation and verification utilities for both ed25519 and secp256r1
   - Implemented test helpers across all contracts for simplified test setup

2. **Security Testing Features**:
   - Implemented cryptographic operations using `ed25519_dalek` library
   - Created verification tools to ensure signatures are properly validated across different algorithms
   - Added authentication flow tests across contract boundaries
   - Implemented tests for device management, account recovery, and factory deployment

3. **Key Test Components**:
   - **AuthController Tests**: Initialization, signer management, threshold enforcement, authentication
   - **Account Tests**: Device management, recovery mechanisms, authentication
   - **AccountFactory Tests**: Account deployment and factory registration
   - **Integration Tests**: Cross-contract interactions and authorization flows
   - **Event Tests**: Verification of proper event emission across all contracts

### ✅ Tests Results
All tests pass successfully, including:
- AuthController tests (initialization, signer management, authentication)
- Account contract tests (device management, recovery)
- AccountFactory tests (deployment flow)
- Integration tests between all three contracts
- Error condition tests for edge cases across the system
- Event emission verification tests for all operations

### Test Coverage
- [x] ✅ Unit Tests
- [x] ✅ Integration Tests
- [ ] 📟 Manual Testing

### Evidence

The test suite verifies functionality across all three contracts:

1. **AuthController Tests**:
   - `test_init`: Validates proper contract initialization with signers and threshold
   - `test_add_signer`: Verifies adding signers with authorization
   - `test_remove_signer`: Ensures signers can be removed properly
   - `test_set_default_threshold`: Validates threshold management
   - `test_add_factory` and `test_add_account`: Tests managing associated contracts
   - Error handling tests for initialization, signer management, and authentication

2. **Account Contract Tests**:
   - `test_add_device`: Tests adding authentication devices
   - `test_remove_device`: Validates device removal functionality
   - `test_add_recovery_address`: Verifies recovery address management
   - Error handling tests for device management and recovery
   - Key security tests for signature validation with secp256r1

3. **AccountFactory Tests**:
   - Tests for account deployment with proper parameters
   - Validation of authentication flow for deployment
   - Event emission verification for deployment events

4. **Integration Tests Across Contracts**:
   - Tests demonstrating AuthController authorizing Account operations
   - Tests showing AccountFactory deploying new Account contracts
   - Full lifecycle tests from factory deployment to account recovery
   - Cross-contract authentication flow validation

5. **Event Testing Across All Contracts**:
   - Verification of events for all operations in each contract
   - Validation of proper event data structures
   - Tests ensuring events provide complete audit trail

### Testing Utilities

The test suite implements specialized utilities for each contract:

1. **AuthController Testing Utilities**:
   ```rust
   pub struct SecureKeyStorage {
       seed: u64,
   }
   
   impl SecureKeyStorage {
       fn new(seed: u64) -> Self { ... }
       fn get_public_key(&self, env: &Env) -> BytesN<32> { ... }
       fn sign(&self, env: &Env, payload: &BytesN<32>) -> BytesN<64> { ... }
       fn get_secp_public_key(&self, env: &Env) -> BytesN<65> { ... }
   }
   ```

2. **Signature generation and verification Testing Utilities**:
   ```rust
   fn create_signed_message(env: &Env, secure_key: &SecureKeyStorage, payload: &BytesN<32>) -> SignedMessage { … }
   fn verify_signature(public_key: &BytesN<32>, payload: &BytesN<32>, signature: &BytesN<64>) -> bool { … }
   fn generate_signers(env: &Env, count: usize) -> Vec<BytesN<32>> { … }
   fn get_secure_key_for_signer(signer_index: usize) -> SecureKeyStorage { … }
   ```

3. **Integration Test Helpers**:
   ```rust
   fn create_auth_client(env: &Env) -> AuthControllerClient { ... }
   fn create_factory_client(env: &Env) -> AccountFactoryClient { ... }
   fn create_account_client(env: &Env) -> AccountContractClient { ... }
   ```

### Comprehensive Test Documentation

A dedicated README.md file was created for the auth-controller tests, providing documentation for all three contracts:

1. AuthController testing documentation
2. Account contract testing overview
3. AccountFactory testing details
4. Integration testing scenarios
5. Instructions for running all tests
6. Troubleshooting guidance for the entire system

The documentation follows a clear structure with sections covering all aspects of the authentication system testing.

### Next Steps
1. Add more comprehensive integration tests simulating real-world usage scenarios, and without mocking auth
2. Expand cryptographic testing to cover additional attack vectors
3. Implement performance testing for computationally intensive operations
4. Add fuzzing tests to discover edge cases
5. Consider adding benchmarking tests for optimization
6. Integrate with CI/CD pipeline for automated test execution

### Conclusion

This PR delivers a comprehensive test suite for the entire KindFi authentication system, covering the AuthController, Account, and AccountFactory contracts. The tests ensure the correctness, security, and reliability of all components individually and as an integrated system. The test suite validates all critical functionality, edge cases, and integration scenarios, providing a strong foundation for the KindFi platform's authentication and security mechanisms. The dedicated documentation makes it easy for developers to understand, run, and extend the tests as the system evolves.

## Summary by CodeRabbit

- **New Features**
  - Added comprehensive test coverage and snapshot files for authentication contract scenarios, including account/device management, signer and factory operations, recovery flows, and event emission.
  - Introduced a detailed README for the authentication controller test suite.
  - Added deployment information for the authentication controller contract.

- **Bug Fixes**
  - Adjusted error code assignments for device and initiation errors in account contract.

- **Refactor**
  - Made `events` modules public in account and account-factory contracts for broader accessibility.
  - Updated event constant names to use uppercase and improved naming consistency.

- **Chores**
  - Updated dependencies and development dependencies in contract configuration files.
  - Removed explicit crate-type declarations from contract configuration files.

### CODE REVIEW

**Comment by psychemist** [comment]
hi @AndlerRL. i'm still working on it. it was taking more time because i was testing all three modules but i'm scaling my tests down to only the auth-contract crate

currently fixing bugs related to my signature generation & verification

i'll add more commits to my PR soon. thanks

**Comment by AndlerRL** [comment]
Hey @psychemist let us know your progress? We want to know how you are rocking! A week has gone and no recent updates provided.

**Comment by psychemist** [comment]
hi @AndlerRL. i'm done with auth controller tests and rounding up on account and factory. i'll have it done by early next week or you can unassign

**Comment by AndlerRL** [comment]
> hi @AndlerRL. i'm done with auth controller tests and rounding up on account and factory. i'll have it done by early next week or you can unassign

Yeah, that's ok @psychemist. If it is early next week, sounds good 🙏  Please keep us updated 👀

**Comment by psychemist** [comment]
@AndlerRL willl be done by EOD

**Comment by AndlerRL** [comment]
> @AndlerRL do i also need to test the check_auth internal function?
> 
> proof of current passing tests: Screenshot 2025-03-25 at 17 26 27

Yes please, that would be the last piece to do test runs :slightly_smiling_face: also, can you add a summary of all the test cases for future reference? :pray:

**Comment by psychemist** [comment]
@AndlerRL got it! will do

**Comment by Bran18** [comment]
Hey @psychemist is this one ready for us to check?

**Comment by psychemist** [comment]
@Bran18 by EOD, it will

**Comment by Bran18** [comment]
> @Bran18 by EOD, it will

Awesome, let us know any update we really want to merge your code

**Comment by psychemist** [comment]
@Bran18 will do. I am currently attempting to enforce auth using the signature verification I implemented. i have been mocking auth for the tests. is this necessary or do i skip it?

**Comment by AndlerRL** [comment]
> @Bran18 will do. I am currently attempting to enforce auth using the signature verification I implemented. i have been mocking auth for the tests. is this necessary or do i skip it?

You could do it @psychemist however, I would prefer to push the latest that you have so far. This test however would be nice to have, so you can leave a code comment about this test case and the intention to it. Due to time-constrains, we can leave it like that and work on it whenever is possible :slightly_smiling_face:

**Comment by psychemist** [comment]
@AndlerRL okay, got it. i'll pause that for now. just figured how to create wasm_hash for account factory so things should pick up from here

**Comment by AndlerRL** [comment]
Sounds very good @psychemist! :+1:

**Comment by psychemist** [comment]
done with factory and account tests. testing check_auth and events now

**Comment by AndlerRL** [comment]
> done with factory and account tests. testing check_auth and events now

Thanks, @psychemist! 😁 Remember that you can let us know once it is done for review but to keep pushing to upstream to keep things up-to date 👀

**Comment by psychemist** [comment]
@AndlerRL @Bran18 PR is ready for review

File: apps/contract/contracts/academy-graduation-nft/src/interface.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+use crate::GraduationNFT;
+use crate::datatype::{NFTError, NFTMetadata};
+use soroban_sdk::{Address, Env, String, Vec};
```
[the remark this answers is not in the document]
Comment (AndlerRL) [reply apps/contract/contracts/academy-graduation-nft/src/interface.rs]: @coderabbitai verify if we have an issue related to this observation or related to the academy NFT contract improvements/bugfixes. I would be good to track this one.

File: apps/contract/contracts/academy-graduation-nft/src/interface.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+/// Trait for handling NFT transfer operations  
+/// For soulbound NFTs, implementations should reject all transfer  
+/// attempts with appropriate error responses.  
+pub trait DistributionOperations {
+    /// Attempts to transfer a token between addresses  
+    ///  
+    /// For soulbound NFTs, this operation will always fail with  
+    /// an appropriate error, as these tokens cannot be transferred.  
+    fn attempt_transfer(from: Address, to: Address, token_id: u128) -> Result<(), NFTError>;
+}
```
[the remark this answers is not in the document]
Comment (AndlerRL) [reply apps/contract/contracts/academy-graduation-nft/src/interface.rs]: @coderabbitai verify if we have an issue related to this observation or related to the academy NFT contract improvements/bugfixes. It would be good to track this one.

File: apps/contract/contracts/academy-graduation-nft/src/datatype.rs [hunk 3]
```diff
@@ -0,0 +1,39 @@
+use soroban_sdk::{Address, String, Vec, contracterror, contracttype};
+
```
[the remark this answers is not in the document]
Comment (AndlerRL) [reply apps/contract/contracts/academy-graduation-nft/src/datatype.rs]: @coderabbitai verify if we have an issue related to this observation (I think we have, a general audit) or related to the academy NFT contract improvements/bugfixes. I would be good to track this one.

**Review by AndlerRL — APPROVED** [review]
LGTM! We finally made it :smiley: test cases looks good, a bit extensive in my opinion to be in a single file. As future improvement, we have to split the test cases by group of scenarios, so readability can highly improve! I could take me much less if we use **modularity** file structure, which turns into a strategic architecture.

Overall, they are good test cases! Thanks for adding the readme summarizing each test case :+1: 

I hope to see you in a next contribution! Of course adhering into our Contribution Guidelines and Code Style and Conventions. These are in our Telegram group and github org. Please [save this](https://kindfis-organization.gitbook.io/development/readme/oss-contribution-guide-how-to-contribute) and [this one](https://kindfis-organization.gitbook.io/development/code-and-design-guide-style-and-conventions/code-and-design-guide-style-and-conventions) for later if you plan to contribute with us in future.

Merging!

File: apps/contract/contracts/academy-graduation-nft/src/datatype.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndlerRL) [reply apps/contract/contracts/academy-graduation-nft/src/datatype.rs]: @coderabbitai no, let's better modify that issue and add this observation to that issue that I am assigned to. This is to reduce issue cluster and since is related, it is OK to handle both on same issue.

**Comment by psychemist** [comment]
@AndlerRL thanks for your patience 😅 i was going to ask if you wanted me to split them into different test suites when i made the PR but it skipped my mind. i look forward to being a future contributor!
